src/waveform/renderers/allshader/digitsrenderer.cpp

```diff
@@ -0,0 +1,194 @@
+#include "waveform/renderers/allshader/digitsrenderer.h"
+
+#include <QColor>
+#include <QFontMetricsF>
+#include <QGraphicsBlurEffect>
+#include <QGraphicsPixmapItem>
+#include <QGraphicsScene>
+#include <QOpenGLTexture>
+#include <QPainter>
+#include <QPainterPath>
+
+#include "waveform/renderers/allshader/matrixforwidgetgeometry.h"
+#include "waveform/renderers/allshader/vertexdata.h"
+
+// Render digits using a texture (generated) with digits with blurred dark outline
+
+namespace {
+
+// The texture will contain 12 characters: 10 digits, colon and dot
+constexpr int NUM_CHARS = 12;
+
+// space around chars for blurred dark outline
+constexpr qreal SPACE = 3;
+// extra kerning when rendering chars
+constexpr qreal KERNING = 1;
+
+char indexToChar(int index) {
+    const char* str = "0123456789:.";
+    return str[index];
+}
+int charToIndex(QChar ch) {
+    int value = ch.digitValue();
+    if (value != -1) {
+        return value;
+    }
+    if (ch == ':') {
+        return 10;
+    }
+    if (ch == '.') {
+        return 11;
+    }
+    assert(false);
+    return 11; // fallback to dot
+}
```

Comment:
digitValue() is unfortunatly not a constexpr, DEBUG_ASSERT will not crash release builds. 

```suggestion
constexpr int charToIndex(QChar ch) {
    int value = ch.toLatin1() - '0';
    if (value >= 0 && value <= 9) {
        return value;
    }
    if (ch == ':') {
        return 10;
    }
    if (ch == '.') {
        return 11;
    }
    DEBUG_ASSERT(false);
    return 11; // fallback to dot
}
```